base/associative.jl

```diff
@@ -136,8 +136,38 @@ This includes arrays, where the keys are the array indices.
 """
 pairs(collection) = Generator(=>, keys(collection), values(collection))
 
+"""
+    similar(a::Associative, [value_type=valtype(a)], [inds=keys(a)])
+
+Create an `Associative` container with unitialized values of `value_type` and indices
+`inds`. The second and third arguments are optional and default to the input's `valtype` and
+`keys`, respectively.
+
+Custom `Associative` subtypes may choose which specific associative type is best suited to
+return for the given value type and indices, by specializing on the three-argument
+signature. The default is to return a `Dict`.
+"""
+similar(a::Associative) = similar(a, valtype(a), keys(a))
+similar(a::Associative, ::Type{T}) where {T} = similar(a, T, keys(a))
+similar(a::Associative, inds) = similar(a, valtype(a), inds)
```

Comment:
`similar(a::Associative, inds::Tuple)` seems like it might solve the ambiguity? You just need to be as specific about the 2nd argument as is https://github.com/JuliaLang/julia/pull/24390/files#diff-2264bb51acec4e7e2219a3cb1c733651R566, and then your specialization on the first argument won't be ambiguous.

Reply:
Thanks, but that part I understand perfectly. :)

I apologize - my post last night was quite unclear (I learned just afterwards that I in-fact had a fever...). What I am saying is that in this work I am imagining a future world where `similar` might have a method like `similar(oldcontainer::Any, newkeys::Any)` return a container similar to `oldcontainer` that can be indexed by the values in `newkeys`, and that `newkeys` could be *any* collection (including a tuple). This directly clashes with https://github.com/JuliaLang/julia/pull/24390/files#diff-2264bb51acec4e7e2219a3cb1c733651R566, and I'm exploring how to make everything work together nicely.

I guess I'm really trying to get my head around what that method helps with? In particular, do you really use `similar(dims->zeros(Int, dims), shape::Tuple)` and how does that differ from `zeros(Int, shape...)`? Would `zeros(eltype, ::AbstractCartesianRange)` cover offset arrays in general? 

I'm also thinking `similar(::AbstractArray, eltype, ::AbstractCartesianRange)` is the "most correct" signature for offset arrays, with some helpers that build Cartesian products when the last argument is varargs and the usual special cases for integers to support the current syntax. That is, basically, have `similar(f, dims::DimOrInd...) = similar(f, to_cartesian_range(dims...))`, and remove the other `Tuple` method entirely. Would that make any sense at all?

Reply:
Sorry to hear about your illness, I hope you're feeling better now!

> In particular, do you really use similar(dims->zeros(Int, dims), shape::Tuple)

Yes, but once we delete [these deprecations](https://github.com/JuliaLang/julia/blob/ccb02de8ccebc76fffe61bdeb823ec181a74edcc/base/deprecated.jl#L1197-L1204) then `zeros(::Type, ::AbstractUnitRange...)` can mean what we need it to mean. And you're right that now that `CartesianRange` wraps an `AbstractUnitRange` tuple (rather than `start`, `stop` `CartesianIndex` "corners" like it did when I introduced this) we could introduce a similar method with `CartesianRange`. So perhaps this method doesn't need to continue to exist.

Reply:
OK, cool, that makes a lot of sense. Thanks. :)

I guess we will deal with this separately then. I actually really wish we could kill those deprecations already (shouldn't that be the logical step zero of developing v0.7?)

Reply:
Generally old deprecations are deleted at the end of the development cycle. But I think in cases of specific need we can justify deleting them early. CC @ararslan for his opinion.

Reply:
I think it depends on when the deprecations were added. If they were pretty recent then it would be good to keep them around longer so that packages have a chance to adjust to the change. But if the deprecations have been around for a version or two, I'd say dump 'em.